### ISSUE TITLE
DropoutWrapper and dynamic_rnn with parallel iterations not reproducible

### ISSUE THREAD

**Issue by alanhdu** [tensorflow/tensorflow#14663]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: CPU Binary (pip wheel)
- **TensorFlow version (use command below)**: v1.4.0-rc1-11-g130a514 1.4.0
- **Python version**: 3.6.3 and 3.5.2
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**:
The following script fails with an assertion error, even though I've explicitly set the random seed:
```python
import tensorflow as tf
import numpy as np

def run():
    tf.reset_default_graph()
    tf.set_random_seed(0)

    sess = tf.Session()
    x = tf.placeholder(tf.float32, [None, None, 1])
    cell = tf.nn.rnn_cell.DropoutWrapper(
        tf.nn.rnn_cell.LSTMCell(100), input_keep_prob=0.5)
    output, state = tf.nn.dynamic_rnn(
        cell, x, dtype=tf.float32, parallel_iterations=100)
    # if parallel_iterations=1, then everything works

    sess.run(tf.global_variables_initializer())
    return sess.run([output, state], {x: np.arange(100).reshape(1, 100, 1)})

o1, (c1, h1) = run()
o2, (c2, h2) = run()
  
assert (o1 == o2).all()
assert (c1 == c2).all()
assert (h1 == h2).all()
```

### Describe the problem
It looks like using parallel iterations options creates some non-determinism when using `DropoutWrapper` (and `parallel_iterations=32` by default). Ideally, when setting the random seed, all TensorFlow operations should be deterministic and reproducible (or the non-determinism should at least be documented).

cc @tudorgt @adamAlnatsheh

**Comment by cy89** [comment]
While I agree with the ideal about deterministic and reproducible, it's not clear to me that we meet that level of guarantee with our current code. @mrry do you have any suggestions?

**Comment by mrry** [comment]
I'm pretty sure that it's impossible to get deterministic results from using an RNG op (like `tf.random_uniform()`) inside a `tf.while_loop()` with `parallel_iterations > 1`, unless there is a data or control dependency between the RNG ops in successive iterations. I can think of two conceptual workarounds, but neither of them is likely to work out of the box with the `tf.rnn.rnn_cell` APIs:

1. Compute the sequence of dropout masks outside the loop using a single `tf.random_uniform()` call, and access the appropriate slice inside the loop.
2. Somehow add control dependencies from the input loop variables (and hence the output values of the previous iteration) to the `tf.random_uniform()` inside `DropoutWrapper`. Technically you only need the control dependency between the `tf.random_uniform()` ops, and the more precise you can make it, the higher the parallelism you'll achieve. (For example, you could add the random mask as a loop variable, purely to make it possible to add a control dependency on it.)

There may be simpler ways to do it, but those are the only approaches I can think of.

/cc @ekelsen for deterministic execution/reproducibility and @ebrevdo for `tf.nn.dynamic_rnn()`.

**Comment by ebrevdo** [comment]
if you want exact reproducibility, use the randomness operations in [`tf.contrib.stateless`](https://www.tensorflow.org/api_docs/python/tf/contrib/stateless).  You'll still have to feed some function of time (i.e., pass it through your state) in order to get proper random seeds for every time step.

I agree we should document nondeterminism of the RNGs when the same op is executed within flow control like while loops but without control dependencies to ensure ordering.

Another question is, can we add some control dependencies in `DropoutWrapper` to ensure ordered execution of all randomness-generating ops.  The answer is maybe, but it's not high priority because not enough folks have considered this to be an issue - and because adding control dependencies generally slows down execution.  I'll leave it to @ekelsen to see if he wants to rifle through the `DropoutWrapper` code to ensure consistency of this sort.

**Comment by ekelsen** [comment]
I think it is unlikely that I'll get to this.  There exists a WAR (use parallel_iterations = 1).  Higher priority is non-determinism in individual ops.

Also to the original poster's comment "when setting the random seed, all TensorFlow operations should be deterministic and reproducible (or the non-determinism should at least be documented)."  This would be a nice state of the world, but far from the world that we live in today.  Many TF operations are not deterministic and there is generally no documentation describing that.  Doing this now for all the ops that exist on the CPU and GPU is a rather daunting task.  I've done work on fixing the most common GPU cases.

**Comment by ebrevdo** [comment]
I think the main remaining source of nondeterminism may be Eigen's float32
matmuls.

**Comment by ahaider3** [comment]
I also came across this issue. I am using MPI_Allreduce to do distributed training of my model. I noticed that my workers were not converging because of this issue. Parallel_iterations == 1 seems to have solved it.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by reedwm** [comment]
@ebrevdo what's the state of this issue? Should it be closed since it seems infeasible to implement? Or marked as contributions welcome?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 106 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 122 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 137 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 152 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 167 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 182 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 197 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 212 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 227 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 242 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 257 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
This is a pretty nontrivial problem to solve.  We can add some dependency constraints to ensure that at least state dropout is performed in a deterministic way, but input dropout and (possibly) output dropout cannot be enforced this way.

Probably the best way to ensure we do deterministic randomness inside a while_loop is to use the random ops from `tf.contrib.stateless` (soon to be moved to core).  This requires a lot more manual management of random seed, and will make the DropoutWrapper more complex.  It's still worth doing, but will require careful thought to do it correctly.

If anyone is interested in implementing a `StatelessDropoutWrapper` using these primitives, I am happy to work with you to make it happen.  cc @girving.

**Comment by ebrevdo** [comment]
cc @qlzh727

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.
